src/tests.rs

```diff
@@ -168,3 +168,408 @@ mod from_items {
+                        original, err,
+                    );
+                }
+
+                format!("unsupported by serde_json: {}", err)
+            }
+        };
+
+        println!("{:?} as JSON: {}", original, original_as_json);
+
+        let actual_serialized = to_attribute_value(original.clone());
+
+        let (expected_serialized_key, actual_serialized) = match expect_serialized_key {
+            Ok(expected_serialized) => (
+                expected_serialized,
+                actual_serialized.unwrap_or_else(|err| {
+                    panic!(
+                        "Failed to serialize to AttributeValue: {}\n\n\
+                    The JSON representation would be:\n{}\n",
+                        err, original_as_json,
+                    )
+                }),
+            ),
+            Err(expected_err) => {
+                assert_eq!(
+                    expected_err,
+                    actual_serialized.expect_err("Expected an error when serializing"),
+                    "Did not get the expected error when serializing"
+                );
+                return;
+            }
+        };
+
+        let expected_serialized = to_attribute_value(hashmap! {
+            expected_serialized_key => "value",
+        })
+        .expect("Failed to serialize expected value");
+        assert_eq!(
+            expected_serialized, actual_serialized,
+            "Serialized value is not what was expected"
+        );
+
+        let deserialized = from_attribute_value(actual_serialized.clone()).unwrap_or_else(|err| {
+            panic!(
+                "Failed to deserialize: {}\nThe serialized value was:\n{:#?}\n",
+                err, actual_serialized
+            )
+        });
+        assert_eq!(
+            original, deserialized,
+            "Deserialized value is not what was expected"
+        );
+    }
+
+    fn key_must_be_a_string<T>() -> Result<T> {
+        Err(ErrorImpl::KeyMustBeAString.into())
+    }
+
+    // Tests using different types of enum variants as map keys.
+    // See: https://serde.rs/enum-representations.html
+    mod enum_variant {
+        use super::*;
+
+        // https://serde.rs/enum-representations.html#externally-tagged
+        mod externally_tagged {
+            use super::*;
+
+            #[derive(
+                Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord, Serialize, Deserialize,
+            )]
+            enum VariantType {
+                Unit,
+                Newtype(String),
+                Struct { value: String },
+                Tuple(String, String),
+            }
+
+            #[test]
+            fn unit_variant() {
+                // TODO: Fix this. It currently panics (`unreachable!()`) in serde_dynamo,
+                //       but works in serde_json. Making it round-trip properly in serde_dynamo
+                //       is more complicated.
+                // map_key_round_trip(VariantType::Unit, Ok("Unit"));
```

Comment:
This is the failing test that's most important to me. The serialization part is easy. Just change `MapKeySerializer::serialize_unit_variant(...)` to this:

```rust

    fn serialize_unit_variant(
        self,
        _name: &'static str,
        _variant_index: u32,
        variant: &'static str,
    ) -> Result<Self::Ok, Self::Error> {
        self.serialize_str(variant)
    }

But this test will still fail because the deserialization doesn't work, and that's not a straightforward fix. I wasn't sure if you might have a quick solution to this one.